### PR TITLE
changing OMDebug based on the new schema

### PR DESCRIPTION
**Pull request by derekpappas**
**Related issue**: 

**Type of change**: feature request

**Impact**: API modification

**Development Phase**:  implementation

**Release Notes**

The OMDebug schema has changed. This PR implements the OMDebug schema changes.

### CODE REVIEW

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 1]
```diff
@@ -31,10 +32,45 @@ class CLINTLogicalTreeNode(device: SimpleDevice, f: => OMRegisterMap) extends Lo
...
+) extends LogicalTreeNode {
   def getOMDebug(resourceBindings: ResourceBindings): Seq[OMComponent] = {
-    val memRegions :Seq[OMMemoryRegion] = DiplomaticObjectModelAddressing.getOMMemoryRegions("Debug", resourceBindings, Some(f))
-    val cfg : DebugModuleParams = debugModuleParams
+    val memRegions: Seq[OMMemoryRegion] = DiplomaticObjectModelAddressing
+      .getOMMemoryRegions("Debug", resourceBindings, Some(f))
+    val cfg: DebugModuleParams = p(DebugModuleParams)
+
+    Seq[OMComponent](
+      OMDebugCSRs(
```
Comment (mwachs5) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: These belong in the Core, not here. Similar to PMPs.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 2]
```diff
@@ -32,45 +33,19 @@ class CLINTLogicalTreeNode(device: SimpleDevice, f: => OMRegisterMap) extends Lo
...
-  }
-}
-
 class DebugLogicalTreeNode(
   device: SimpleDevice,
-  f: => OMRegisterMap,
-  p: Parameters,
-  m: () => TLDebugModule
-) extends LogicalTreeNode {
+  dmOuter: () => TLDebugModuleOuterAsync,
```
Comment (mwachs5) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: why do you want the `Async` versions of these? Why not just pass in dmOuter.dmOuter in the calling function?

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 3]
```diff
@@ -32,45 +33,19 @@ class CLINTLogicalTreeNode(device: SimpleDevice, f: => OMRegisterMap) extends Lo
...
+  dmInner: () => TLDebugModuleInnerAsync
+)(implicit val p: Parameters) extends LogicalTreeNode {
   def getOMDebug(resourceBindings: ResourceBindings): Seq[OMComponent] = {
+    val nComponents: Int = dmOuter().dmOuter.module.getNComponents()
```
Comment (mwachs5) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: Above comment would mean you don't need to keep calling the nested versions here.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala (hunk 2 above)
Comment (derekpappas) [reply src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: There is an eval order problem. I should have saved the exception message.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala (hunk 3 above)
Comment (derekpappas) [reply src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: This code is not executed until after the test harness memories are instantiated and control returns to the EnterpriseGenerator. @terpstra knows about this problem

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 4]
```diff
@@ -46,10 +57,41 @@ class DebugLogicalTreeNode(device: SimpleDevice, f: => OMRegisterMap, debugModul
...
+        hartSeltoHartIDMapping = Nil, // HartSel goes from 0->N but HartID is not contiguious or increasing
+        authenticationType = NONE,
+        nHartsellenBits = p(MaxHartIdBits), // Number of actually implemented bits of Hartsel
+        hasHartInfo = true,
+        hasAbstractauto = true,
+        cfgStrPtrValid = false,
+        nHaltSummaryRegisters = 2,
+        nHaltGroups = cfg.nHaltGroups,
+        nExtTriggers = cfg.nExtTriggers,
+        hasResetHaltReq = true,
```
Comment (ernie-sifive) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: Hi Derek,
A couple of these fixed-value items do not have the right values for the current DM implementation.  They should be:
hasResetHaltReq = false
hasAbstractPostExec = true

Comment (derekpappas) [reply src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: Hi Ernie,

Please make a PR to fix the values.

Comment (ernie-sifive) [reply src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: https://github.com/freechipsproject/rocket-chip/pull/1938

**Pull request merged**